pandas/_libs/src/klib/khash_python.h

```diff
@@ -351,6 +393,10 @@ khuint32_t PANDAS_INLINE kh_python_hash_func(PyObject* key) {
     else if (PyTuple_CheckExact(key)) {
         hash = tupleobject_hash((PyTupleObject*)key);
     }
+    else if (PyObject_TypeCheck(key, &PyDatetimeArrType_Type)) {
+        // GH#50690
+        hash = np_datetime64_object_hash(key);
```

Comment:
```suggestion
        hash = np_datetime64_object_hash((PyDatetimeScalarObject *)key);
```
As said above, it's just nice to do the cast where you check.